### ISSUE TITLE
Implement VMAF support

### ISSUE THREAD

**Issue by shssoichiro** [rust-av/av-metrics#14]
The codebase for VMAF is quite large, but it would be nice to have a pure Rust implementation. This is by no means high priority, just if someone (probably me) gets bored.

**Comment by colbyn** [comment]
Ahh yes!

I’m working on an image optimization service for the e-commerce world (if anyones interested please do contact me!) that uses VMAF internally in scheme similar to a brute force search. 

I’ve been using it VIA the C API for some time, and while I love VMAF, I still haven’t yet figured out what `temp_data` does (it’s not used in the FFmpeg implementation):

```C
int compute_vmaf(double* vmaf_score, char* fmt, int width, int height, 
int (*read_frame)(float *ref_data, float *main_data, float *temp_data, int stride, void *user_data), 
void *user_data, char *model_path, char *log_path, char *log_fmt, int disable_clip, 
int disable_avx, int enable_transform, int phone_model, int do_psnr, int do_ssim, 
int do_ms_ssim, char *pool_method, int thread, int subsample, int enable_conf_interval);
```

The callback argument is weird because I can’t fill the buffers directly with a single frame and then return the value 2 (what seems to signal to VMAF that it’s done). 


Anyway what I wanted to mention. I still haven’t figured out why and perhaps it’s me, but for some reason I can’t run multiple instances of VMAF in a single process in parallel on different media. It’s significantly more likely to segfault. If it’s a bug with VMAF it’s also not deterministic for some reason; sometimes it runs just fine…

**Comment by colbyn** [comment]
Also for anyone interested, I just publish a VMAF wrapper to crates here [vmaf-sys](https://crates.io/crates/vmaf-sys). 

It’s pretty direct. But it does the grunt work of downloading, building and statically linking `libvmaf` internally (so you don’t have to have it pre-installed on your system, like when building FFmpeg). 

Also the default and 4K model is baked into the binary.

**Comment by shssoichiro** [comment]
Thanks @colbyn! I think I will go ahead and use vmaf-sys to get basic support in this crate, then Rewrite it In Rust at a later time.

**Comment by colbyn** [comment]
@shssoichiro Okay sounds good. Let me know if you have any problems.

**Comment by shssoichiro** [comment]
I'm having some difficulties with the wrapper you created, and wondering if there's a good way to resolve this via Rust FFI. It looks like the run function takes a `model_path` which needs to point to a path representing one of the model files, but that's not possible the way the wrapper is currently set up--pulling the models in with `include_str`.

**Comment by colbyn** [comment]
@shssoichiro Oh I was just working on this in my own new implementation.

I never realized this until just now. The `compute_vmaf ` function has an argument called `model_path` that also at runtime checks for a second file with the same but but with a `.model` extension (I never realized this since before I just included the whole models dir from the VMAF git repo).

Looks like the docs hasn't built yet. But for the `0.0.8` version you can see a helper function [here](https://github.com/colbyn/vmaf-sys/blob/81ac78cc60cc42db94b5fcad42421210f6f522e6/src/extras.rs#L156) that should give you a valid file path reference that you can give to VMAF that also includes the secondary (`*.pkl.model`) file.

```rust
use std::path::PathBuf;
use std::ffi::CStr;
use std::ffi::CString;
use std::os::raw::c_char;
use std::os::raw::c_int;
use libc::{size_t, c_float};

// ...

// SETTINGS
let mut vmaf_score = 0.0;
let model_path = vmaf_sys::extras::get_def_model_path()
    .to_str()
    .expect("PathBuf to str failed")
    .to_owned();
let model_path = CString::new(model_path).expect("CString::new failed");
let mut fmt = CString::new(String::from("yuv420p")).expect("CString::new failed");
let width = source1.width;
let height = source1.height;
let log_path: *mut c_char = std::ptr::null_mut();
let log_fmt: *mut c_char = std::ptr::null_mut();
let disable_clip = 0;
let disable_avx = 0;
let enable_transform = 0;
let phone_model = 0;
let do_psnr = 0;
let do_ssim = 0;
let do_ms_ssim = 0;
let pool_method: *mut c_char = std::ptr::null_mut();
let n_thread = 1;
let n_subsample = 1;
let enable_conf_interval = 0;

// GO!
let compute_vmaf_res = vmaf_sys::compute_vmaf(
    &mut vmaf_score,
    fmt.as_ptr() as *mut c_char,
    width as c_int,
    height as c_int,
    Some(read_frame),
    vmaf_ctx as *mut libc::c_void,
    model_path.as_ptr() as *mut c_char,
    log_path,
    log_fmt,
    disable_clip,
    disable_avx,
    enable_transform,
    phone_model,
    do_psnr,
    do_ssim,
    do_ms_ssim,
    pool_method,
    n_thread,
    n_subsample,
    enable_conf_interval
);
```

**Comment by colbyn** [comment]
@shssoichiro Did that solve your problem?

**Comment by shssoichiro** [comment]
Yes, thanks! Now I just have to figure out how this read_frame thing works... :slightly_smiling_face: :

**Comment by colbyn** [comment]
Ahh well if it helps, here is the snippet I’m using:
```rust
#[repr(C)]
struct VmafReportContext {
    source1: DecodedImage,
    source2: DecodedImage,
    frame_set: bool,
}

unsafe extern "C" fn read_frame(
    mut source1_out: *mut c_float,
    mut source2_out: *mut c_float,
    temp_data: *mut c_float,
    out_stride: c_int,
    raw_ctx: *mut libc::c_void,
) -> c_int {
    // CONTEXT
    let mut vmaf_ctx = Box::from_raw(raw_ctx as *mut VmafReportContext);
    let mut vmaf_ctx = Box::leak(vmaf_ctx);

    // RESOLUTION
    let width = vmaf_ctx.source1.width;
    let height = vmaf_ctx.source1.height;

    // Y PLANE DATA
    let mut source1_in: *mut u8 = vmaf_ctx.source1.buffer.as_mut_ptr();
    let mut source2_in: *mut u8 = vmaf_ctx.source2.buffer.as_mut_ptr();

    // Y PLANE LINESIZE
    let source1_linesize: usize = vmaf_ctx.source1.linesize as usize;
    let source2_linesize: usize = vmaf_ctx.source2.linesize as usize;
    
    // OUTPUT LINESIZE
    let out_stride = out_stride as usize;

    // DONE
    if vmaf_ctx.frame_set {
        return 2;
    }
    
    // FILL BUFFERS (THE EXTREMELY UNSAFE, DANGEROUS AND CONFUSING PART)
    for y in 0..height {
        for x in 0..width {
            // GET - SOURCE 1 & 2
            let s1_px: u8 = *(source1_in.offset(x as isize));
            let s2_px: u8 = *(source2_in.offset(x as isize));

            // CONVERT - SOURCE 1 & 2
            let s1_px: c_float = s1_px as c_float;
            let s2_px: c_float = s2_px as c_float;

            // SET - OUTPUT 1 & 2
            *(source1_out.offset(x as isize)) = s1_px;
            *(source2_out.offset(x as isize)) = s2_px;
        }
        // UPDATE - SOURCE 1
        source1_in = source1_in.add(source1_linesize / std::mem::size_of_val(&*source1_in));
        source1_out = source1_out.add(out_stride / std::mem::size_of_val(&*source1_out));

        // UPDATE - SOURCE 2
        source2_in = source2_in.add(source2_linesize / std::mem::size_of_val(&*source2_in));
        source2_out = source2_out.add(out_stride / std::mem::size_of_val(&*source2_out));
    }
    vmaf_ctx.frame_set = true;
    return 0;
}
```
But this is written for a single frame, so perhaps not so useful for your case. 

Also this is the locking snippet I’m using (maybe it could be better, idk):
```rust
lazy_static! {
    static ref VMAF_LOCK: Mutex<()> = {
        Mutex::new(())
    };
}

// …

// LOCK
let lock = VMAF_LOCK.lock().expect("failed to lock vmaf work");
// GO!
let score = unsafe {
    vmaf_controller(source1.clone(), source2.clone())
};
// UNLOCK
std::mem::drop(lock); // Just to be explicit
```

Regarding the `user_data` argument, I'm initializing the data (in my case a struct called `VmafReportContext`) like this (before passing such to `read_frame`):
```rust
// INIT VMAF CONTEXT
let mut vmaf_ctx = Box::new(VmafReportContext {
    source1: source1.clone(),
    source2: source2.clone(),
    frame_set: false
});
let vmaf_ctx = Box::into_raw(vmaf_ctx);

// ... call `compute_vmaf`...

// CLEANUP
let mut vmaf_ctx = Box::from_raw(vmaf_ctx);
std::mem::drop(vmaf_ctx);  // Just to be explicit
```

But heres the [FFmpeg implementation](https://github.com/FFmpeg/FFmpeg/blob/e835a9d30263ce16e2621d79ff1be9afd5274a60/libavfilter/vf_libvmaf.c#L92) which is probably closer to what you’re looking for.

PS: For me, writing to the `ref_data` and `main_data` buffers is the hard part since I’m not very comfortable with is pixel format conversions. It’ll be interesting how `rust-av` supports this :)
